File: web/src/lib/services/timerService.ts
import { CookingTimer } from '@/types/models'

class TimerService {
  private timers: Map<string, NodeJS.Timeout> = new Map()
  private callbacks: Map<string, (timer: CookingTimer) => void> = new Map()

  startTimer(timer: CookingTimer, onTick: (timer: CookingTimer) => void, onComplete: () => void) {
    this.stopTimer(timer.id)
    
    this.callbacks.set(timer.id, onTick)
    
    const interval = setInterval(() => {
      timer.remainingTime -= 1
      
      if (timer.remainingTime <= 0) {
        this.stopTimer(timer.id)
        onComplete()
      } else {
        onTick(timer)
      }
    }, 1000)
    
    this.timers.set(timer.id, interval)
  }

  stopTimer(timerId: string) {
    const interval = this.timers.get(timerId)
    if (interval) {
      clearInterval(interval)
      this.timers.delete(timerId)
      this.callbacks.delete(timerId)
    }
  }

  pauseTimer(timerId: string) {
    this.stopTimer(timerId)
  }

  stopAllTimers() {
    this.timers.forEach((_, id) => this.stopTimer(id))
  }
}

export const timerService = new TimerService()

File: android/app/src/main/java/com/neostudios/zest/data/local/TimerService.kt
package com.neostudios.zest.data.local

import android.content.Context
import android.os.CountDownTimer
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

data class CookingTimer(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val durationSeconds: Int,
    val remainingSeconds: Int,
    val isRunning: Boolean = false,
    val recipeId: String? = null
)

@Singleton
class TimerService @Inject constructor(
    @ApplicationContext private val context: Context
) {
    private val _timers = MutableStateFlow<List<CookingTimer>>(emptyList())
    val timers: StateFlow<List<CookingTimer>> = _timers.asStateFlow()
    
    private val activeTimers = mutableMapOf<String, CountDownTimer>()

    fun addTimer(name: String, durationSeconds: Int, recipeId: String? = null) {
        val timer = CookingTimer(
            name = name,
            durationSeconds = durationSeconds,
            remainingSeconds = durationSeconds,
            recipeId = recipeId
        )
        _timers.value = _timers.value + timer
    }

    fun startTimer(timerId: String) {
        val timer = _timers.value.find { it.id == timerId } ?: return
        if (timer.isRunning) return

        val countDownTimer = object : CountDownTimer(
            timer.remainingSeconds * 1000L,
            1000
        ) {
            override fun onTick(millisUntilFinished: Long) {
                val remainingSeconds = (millisUntilFinished / 1000).toInt()
                updateTimer(timerId) { it.copy(remainingSeconds = remainingSeconds) }
            }

            override fun onFinish() {
                updateTimer(timerId) { it.copy(remainingSeconds = 0, isRunning = false) }
                NotificationHelper.sendTimerComplete(context, timer.name)
                activeTimers.remove(timerId)
            }
        }

        countDownTimer.start()
        activeTimers[timerId] = countDownTimer
        updateTimer(timerId) { it.copy(isRunning = true) }
    }

    fun pauseTimer(timerId: String) {
        activeTimers[timerId]?.cancel()
        activeTimers.remove(timerId)
        updateTimer(timerId) { it.copy(isRunning = false) }
    }

    fun resetTimer(timerId: String) {
        activeTimers[timerId]?.cancel()
        activeTimers.remove(timerId)
        updateTimer(timerId) { 
            it.copy(
                remainingSeconds = it.durationSeconds,
                isRunning = false
            )
        }
    }

    fun deleteTimer(timerId: String) {
        activeTimers[timerId]?.cancel()
        activeTimers.remove(timerId)
        _timers.value = _timers.value.filter { it.id != timerId }
    }

    private fun updateTimer(timerId: String, update: (CookingTimer) -> CookingTimer) {
        _timers.value = _timers.value.map { timer ->
            if (timer.id == timerId) update(timer) else timer
        }
    }

    fun cleanup() {
        activeTimers.values.forEach { it.cancel() }
        activeTimers.clear()
    }
}
